src/cudadecoder/cuda-pipeline-common.h

```diff
@@ -100,13 +102,13 @@ class CudaPipelineResult {
     ctm_result_ = std::move(ctm);
   }
 
-  const CompactLattice &GetLatticeResult() const {
+  CompactLattice &GetLatticeResult() {
     KALDI_ASSERT("Lattice result was not requested" &&
                  result_type_ & RESULT_TYPE_LATTICE);
     return clat_;
   }
 
-  const CTMResult &GetCTMResult() const {
+  CTMResult &GetCTMResult() {
```

Comment:
Switching to pure C pointers. I guess the official C++11 way would be using a shared_ptr, but this triggers atomics, and moving unique_ptrs seems like a hack in that case. Do you think we can just use C pointers ?

Reply:
Absolutely, this is exactly what Google code style prescribes! Before C++11, members like `X* x_;` had to be commented as to whether the class owns or does not own the object. With the C++11 `unique_ptr`, the new convention is to use the former to indicate non-owned pointer, the latter owned. This eliminates all explicit `new` and `delete` in higher-level, problem domain classes (the running joke is `make_unique` is the new `new`). The style places a lot of weight on _conventions_: of many ways something can be written equivalently, they try to assign semantics to a specific syntax. Kaldi codebase is older than these conventions, so you'll find many `new`/`delete`s.

For passing things around, raw pointers are A-OK and encouraged, both const and non-const. The [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) is a document really worth reading, not simply using as a reference. It does not simply says "this is allowed, this is not;" every section discusses cons and pros of a pattern, and is concluded with a decision, which is not always binary (e.g., discouraged but allowed). It's very reasonable in the sense of using language features where they belong. The prohibition of mutable references is there to make mutability very explicit in code, even if you do not see the definition: `.` vs `->`. 

The Google's prescription for a subobject accessor is:
```c++
Type* mutable_subobject();
const Type& subobject() const;
// const Type* subobject() const;  // is used instead if `nullptr` has a defined semantics
```
[Kaldi deviates in always using PascalCase](http://kaldi-asr.org/doc/style.html) vs. Google asking for snake_case for primitive single-liners, like accessors, size(), etc., and PascalCase for higher level methods) use `MutableSubobject` and `GetSubobject` names instead.

Smart pointers are used only for smart things, like the indication of whether a pointer is owned or not by an object. If a method returns unique_ptr, it is assumed that it hands over the ownership to the caller (it's a syntactically transparent convention as RVO eliminates `move()` to express move semantics); you just _know_ that iff you hold the unique_ptr, you're the responsible sole owner, otherwise not.

The Google style evolves with the language, naturally. It used to be very conservative, so that new engineers could easily read the code even switching from other language background (Google's interview does not require knowledge of any language in particular, and they do not hire to a specific position). For example, one of my interviewers wanted a design of a Java class, and we went through it in C#, the closest language I knew. At that time (2012) move semantics was not yet allowed, and rightly so: C++14 is joked to be "C++11 Service Pack 1", and fixed a lot of defects in this area. E.g., it was unspecified if compiler should generate a copy constructor if the move constructor is explicitly deleted—and Visual Studio 2013 compiler thought no, but 2015 switched to yes. It's still undefined in C++11, but all compilers sensibly "backport" the decision from C++14 into C++11 (which is "yes"), so C++14 seriously strengthened portability of C++11 in this sense.